src/storage/storage-types.ts

```diff
@@ -26,14 +22,71 @@ import {
 } from '../util/errors';
 
 //================================================================================
+// TYPES AND EVENTS
 
 export type StorageId = string;
 
-export type StorageEvent =
-    'ingest' |  // 'ingest|/some/path.txt'
-    'willClose' | 'didClose';
+export type StorageBusChannel =
+    'ingest' |  // 'write|/some/path.txt'  // note that write errors and no-ops are also sent here
+    'willClose' |
+    'didClose';
+
+export interface QueryResult {
+    // the docs from the query...
+    docs: Doc[],
+    // ...and the storageDriver's maxLocalIndex at the time
+    // just before and just after the query was done.
+    // This provided a lower and upper bound for the maxLocalIndex
+    // associated with the resulting docs.
+    // (This is the OVERALL max local index for
+    // the whole storage, not just for the resulting docs.)
+    maxLocalIndexBefore: number,
+    maxLocalIndexAfter: number,
+    // The max localIndex out of the returned docs.
+    // This could be much smaller than the overall maxLocalIndex
+    // if the docs have been filtered.
+    // If there are no matching docs, this is -1.
+    maxLocalIndexInResult: number,
+}
```

Comment:
Now when we do a query we get back a richer result.  Read comments above for details.

It gives us an upper and lower bound on changes that might have happened during the query, so we know if the query is complete or not.